File: finanze/infrastructure/client/rates/exchange_rate_client.py
import logging
from datetime import datetime

import requests
from application.ports.exchange_rate_provider import ExchangeRateProvider
from cachetools import TTLCache, cached
from domain.dezimal import Dezimal
from domain.exchange_rate import ExchangeRates

AVAILABLE_CURRENCIES = ["EUR", "USD"]


def _parse_rates(rates: dict) -> dict:
    return {currency.upper(): Dezimal(rate) for currency, rate in rates.items()}


class ExchangeRateClient(ExchangeRateProvider):
    MATRIX_CACHE_TTL = 2 * 60 * 60
    TIMEOUT = 10

    BASE_URL = "https://cdn.jsdelivr.net/npm/@fawazahmed0/currency-api@latest/v1"
    CURRENCIES_URL = f"{BASE_URL}/currencies.min.json"
    DATE_FORMAT = "%Y-%m-%d"

    def __init__(self):
        self._rates = {}
        self._available_currencies = {}
        self._update_date = None
        self._log = logging.getLogger(__name__)

        self._load_rate_matrix(self.TIMEOUT)

    def get_available_currencies(self, **kwargs) -> dict[str, str]:
        if not self._available_currencies:
            timeout = kwargs.get("timeout", self.TIMEOUT)
            self._available_currencies = self._fetch_available_currencies(timeout)
        return self._available_currencies

    @cached(cache=TTLCache(maxsize=1, ttl=MATRIX_CACHE_TTL))
    def get_matrix(self, **kwargs) -> ExchangeRates:
        current_date = self._get_current_date()
        timeout = kwargs.get("timeout", self.TIMEOUT)
        if current_date != self._update_date:
            self._load_rate_matrix(timeout)
        return self._rates

    def _fetch_available_currencies(self, timeout: int) -> dict:
        return self._fetch(self.CURRENCIES_URL, timeout)

    def _fetch_rates(self, currency: str, timeout: int) -> dict:
        url = f"{self.BASE_URL}/currencies/{currency.lower()}.min.json"
        return self._fetch(url, timeout)

    def _fetch(self, url: str, timeout: int) -> dict:
        response = requests.get(url, timeout=timeout)
        if response.ok:
            return response.json()

        self._log.error("Error Response Body:" + response.text)
        response.raise_for_status()
        return {}

    def _get_current_date(self) -> str:
        return datetime.now().strftime(self.DATE_FORMAT)

    def _load_rate_matrix(self, timeout: int):
        for currency in AVAILABLE_CURRENCIES:
            result = self._fetch_rates(currency, timeout=timeout)
            self._update_date = datetime.strptime(result["date"], self.DATE_FORMAT)
            self._rates[currency] = _parse_rates(result[currency.lower()])

File: mobile/src/services/client/rates/exchangeRateClient.ts
import { ExchangeRateProvider } from "@/application/ports"
import { Dezimal } from "@/domain/dezimal"
import { ExchangeRates } from "@/domain"
import { httpGetJson } from "@/services/client/http/httpClient"

const AVAILABLE_CURRENCIES = ["EUR", "USD"]

const parseRates = (
  rates: Record<string, unknown>,
): Record<string, Dezimal> => {
  const parsed: Record<string, Dezimal> = {}
  for (const [currency, rate] of Object.entries(rates)) {
    parsed[currency.toUpperCase()] = Dezimal.fromString(String(rate))
  }
  return parsed
}

export class ExchangeRateClient implements ExchangeRateProvider {
  static readonly MATRIX_CACHE_TTL_MS = 2 * 60 * 60 * 1000
  static readonly TIMEOUT = 10

  static readonly BASE_URL =
    "https://cdn.jsdelivr.net/npm/@fawazahmed0/currency-api@latest/v1"
  static readonly CURRENCIES_URL = `${ExchangeRateClient.BASE_URL}/currencies.min.json`

  private rates: ExchangeRates = {}
  private availableCurrencies: Record<string, string> = {}
  private updateDate: Date | null = null

  private matrixCacheExpiresAt: number | null = null
  private matrixLoadPromise: Promise<void> | null = null

  constructor() {
    // Mirror backend behavior: eagerly load once on init.
    this.matrixLoadPromise = this.loadRateMatrix(
      ExchangeRateClient.TIMEOUT,
    ).catch(e => {
      console.warn("Failed to load initial rate matrix", e)
    })
  }

  async getAvailableCurrencies(kwargs: any): Promise<Record<string, string>> {
    if (!Object.keys(this.availableCurrencies).length) {
      const timeout = kwargs?.timeout ?? ExchangeRateClient.TIMEOUT
      this.availableCurrencies = await this.fetchAvailableCurrencies(timeout)
    }
    return this.availableCurrencies
  }

  async getMatrix(kwargs: any): Promise<ExchangeRates> {
    const now = Date.now()

    if (this.matrixCacheExpiresAt !== null && now < this.matrixCacheExpiresAt) {
      return this.rates
    }

    if (!this.matrixLoadPromise) {
      const timeout = kwargs?.timeout ?? ExchangeRateClient.TIMEOUT

      // Mirror backend logic: reload when cache misses and date "changes".
      // Note: backend compares a date-string to a datetime, which is always != once initialized.
      const currentDate = this.getCurrentDate()
      if (currentDate !== (this.updateDate as any)) {
        this.matrixLoadPromise = this.loadRateMatrix(timeout)
      } else {
        this.matrixLoadPromise = this.loadRateMatrix(timeout)
      }
    }

    await this.matrixLoadPromise.finally(() => {
      this.matrixLoadPromise = null
    })

    this.matrixCacheExpiresAt =
      Date.now() + ExchangeRateClient.MATRIX_CACHE_TTL_MS
    return this.rates
  }

  private async fetchAvailableCurrencies(
    timeout: number,
  ): Promise<Record<string, string>> {
    const { response, data } = await httpGetJson<Record<string, string>>(
      ExchangeRateClient.CURRENCIES_URL,
      undefined,
      timeout,
    )

    if (!response.ok) {
      console.error("Error Response Body:" + JSON.stringify(data))
      throw new Error(
        `Failed to fetch available currencies: ${response.status}`,
      )
    }

    return data
  }

  private async fetchRates(currency: string, timeout: number): Promise<any> {
    const url = `${ExchangeRateClient.BASE_URL}/currencies/${currency.toLowerCase()}.min.json`
    const { response, data } = await httpGetJson<any>(url, undefined, timeout)

    if (!response.ok) {
      console.error("Error Response Body:" + JSON.stringify(data))
      throw new Error(
        `Failed to fetch rates for ${currency}: ${response.status}`,
      )
    }

    return data
  }

  private getCurrentDate(): string {
    const d = new Date()
    const yyyy = String(d.getFullYear())
    const mm = String(d.getMonth() + 1).padStart(2, "0")
    const dd = String(d.getDate()).padStart(2, "0")
    return `${yyyy}-${mm}-${dd}`
  }

  private async loadRateMatrix(timeout: number): Promise<void> {
    for (const currency of AVAILABLE_CURRENCIES) {
      const result = await this.fetchRates(currency, timeout)
      const dateStr = result?.date
      if (typeof dateStr === "string") {
        this.updateDate = new Date(dateStr)
      }
      const lower = currency.toLowerCase()
      const rawRates = result?.[lower] ?? {}
      this.rates[currency] = parseRates(rawRates)
    }
  }
}
